### ISSUE TITLE
What to return from arbcall?

### ISSUE THREAD

**Issue by kalmarek** [kalmarek/Arblib.jl#16]
this is just to have a space for discussion on the topic rised in [here](https://github.com/kalmarek/Arblib.jl/pull/15#issuecomment-616193596):
> @Joel-Dahne wrote:
I think they case of `Bool` versus `Int32` is interesting as well. While many of the int types in arb represents booleans, some of them are flags with more possible values. It would fit much better in Julia to return a `Bool` instead of an `Int32` when possible, but then we need some way of determining when this can be done. When it occurs as an argument we can use `Union{Bool, Int32}` or maybe even `Integer`.

Another solution is to specialize `@arbcall macro on classes of functions, (one of them: predicates), e.g. 
```julia
@arbcall "arb_add(...)" # normal function
@arbcall Predicate "int arb_is_zero(const arb_t x)" iszero
```
which should call
```julia
macro arbcall(ftype::Expr, arbsignature_str::String,
    jl_fname=jlfname(Arbfunction(arbsignature_str)))
    af = Arbfunction(arbsignature_str)
    local T = eval(ftype)
    return jlcode(T, af, jl_fname)
end

function jlcode(::Type{Predicate}, af::Arbfunction, jl_fname=jlfname(af))
    returnT = returntype(af)
    cargs = arguments(af)
    args, kwargs = jlargs(af)

    call = :(
        function $jl_fname($(args...); $(kwargs...))
              res = ccall(Arblib.@libarb($(arbfname(af))),
              $returnT,
              $(Expr(:tuple, ctype.(cargs)...)),
              $(Symbol.(name.(cargs))...))
              return !iszero(res)
        end
    )
end
```
(I don't like this `eval`, but i'm not sure what to do about it)

**Comment by kalmarek** [comment]
one potential idea is to have a look at jump macros to see how do they handle such constructs:
```julia
m = Model()
@variable m x[1:5]
```

**Comment by Joel-Dahne** [comment]
You are referring to the Julia JuMP package? I'm not entirely sure what you mean, what is it that we should see how they handle?

**Comment by kalmarek** [comment]
yes, the JuMP package; 
they have a set of macros: `@variable`, `@constraint`, `@objective` that take as first argument an object (a model) and dispatch somehow on its type (e.g. you can have `PolyModel()` for polynomial opimization, etc.). 
But anyway, maybe it's the time to test our macros in practice and start producing actual working functions ;)

**Comment by Joel-Dahne** [comment]
I see! But I agree, we can wait with this detail (which should be somewhat easy to change later) and try to get some actual working function!

**Comment by saschatimme** [comment]
Now that I am working a little bit with the library, it is super annoying that most inplace functions just return `nothing`. What do you think about returning the first argument if the return value is `void` anyways ?

**Comment by saschatimme** [comment]
Also could we maybe consider a more manual appraoch for all the functions which return `int`. There are plenty of functions where the `int` is actually just representing a `Bool`. Maybe some manual overrides?

**Comment by kalmarek** [comment]
I'm not sure about automatic conversion to Bool since this needs to be done on case by case basis. This is essentially what we do in `predicates.jl`:
```julia
    (Acb, (
        (:isexact, :is_exact),
        (:(Base.isfinite), :is_finite),
        (:(Base.isinteger), :is_int),
        (:(Base.isone), :is_one),
        (:(Base.isreal), :is_real),
        (:(Base.iszero), :is_zero),
    ))
    )
    for (jlf, arbf) in funcpairs
        @eval $jlf(x::$T) = !iszero($arbf(x))
    end
```

**Comment by saschatimme** [comment]
Oh yes, I didn't want to say that we should do this automatically. But is there maybe a more lightweight approach?

**Comment by kalmarek** [comment]
resolved by #37 ?
